plugins/techdocs/src/client.ts

```diff
@@ -51,6 +51,70 @@ export class TechDocsClient implements TechDocsApi {
     this.fetchApi = options.fetchApi;
   }
 
+  private async getCookie(): Promise<{ expiresAt: string }> {
+    const apiOrigin = await this.getApiOrigin();
+    const requestUrl = `${apiOrigin}/cookie`;
+    const response = await this.fetchApi.fetch(`${requestUrl}`, {
+      credentials: 'include',
+    });
+    if (!response.ok) {
+      throw await ResponseError.fromResponse(response);
+    }
+    return await response.json();
+  }
+
+  private async refreshCookie(options: {
+    expiresAt?: Date | null;
+    timeoutId: number | null;
+  }): Promise<{ expiresAt: string }> {
+    // Always clear the previous timeout
+    if (options.timeoutId) clearTimeout(options.timeoutId);
+
+    // Only issue a new cookie when we don't have an expiry date or it's in the past
+    if (!options.expiresAt || options.expiresAt.getTime() < Date.now()) {
+      return this.getCookie().then(response => {
+        const expiresAt = new Date(response.expiresAt);
+        expiresAt.setMinutes(expiresAt.getMinutes() - 5);
+        localStorage.setItem(
+          'backstage-auth-cookie-last-refresh-expiration-date',
+          expiresAt.toISOString(),
+        );
+        const timeoutId = setTimeout(this.refreshCookie, expiresAt.getTime());
+        localStorage.setItem(
```

Comment:
mmm it doesn't make so much sense to persist `timeoutId` in localStorage as `timeoutId` will be invalid on page reload. What's the use case?

Reply:
When people navigate to another plugin and come back to the techdocs reader, I thought we would have to cancel the previous timeout because we are not creating the timeout in a useEffect and canceling it there... Does this make sense to you?

Reply:
ah I see! Then maybe it could make sense to store the `timeoutId` in the client itself? Otherwise we might get conflicts in case Backstage is opened on multiple tabs

Reply:
The cookie persists across reloads and across tabs though right? And the timeout was intended to reflect that, I presume.

Best to validate that of course. And maybe it can be argued that it's best to fetch it once too many times at least on new tabs etc.

Reply:
Thank you for the inputs @vinzscam and @freben , I tried to reimplement this, could you take a look again 🙏🏻 ?

Reply:
@Rugvip helped me to refactor this in a React hook form 🙏🏻